File: bestfitpy/tests/test_stats.py
"""Stats utility wrapper tests.

Numeric expectations come from the data_utility fixture
(fixtures/data/statistics_utilities.json) routed through the PUBLIC wrappers; the
rest are structural (round trips, shapes, argument validation) with no oracle
content.
"""

import json
from importlib.resources import files
from pathlib import Path

import numpy as np
import pytest

import bestfitpy as bf


def _fixtures_dir() -> Path:
    try:
        packaged = files("bestfitpy") / "fixtures"
        if packaged.is_dir():
            return Path(str(packaged))
    except (ModuleNotFoundError, FileNotFoundError):
        pass
    return Path(__file__).resolve().parents[2] / "fixtures"


def test_public_stats_reproduce_data_utility_oracles():
    spec = json.loads((_fixtures_dir() / "data" / "statistics_utilities.json").read_text())
    datasets = {k: list(v) for k, v in spec["datasets"].items()}
    for case in spec["cases"]:
        fn = case["function"]
        args = case.get("args", [])
        data = datasets.get(case.get("dataset"), [])
        if fn == "MGBT":
            actual = float(bf.mgbt_test(data))
        elif fn == "BoxCoxLambda":
            actual = bf.box_cox_lambda(data)
        elif fn == "BoxCoxTransform":
            actual = bf.box_cox(data, args[0])[int(args[1])]
        elif fn == "YeoJohnsonLambda":
            actual = bf.yeo_johnson_lambda(data)
        elif fn == "YeoJohnsonTransform":
            actual = bf.yeo_johnson(data, args[0])[int(args[1])]
        elif fn == "PlottingPosition":
            actual = bf.plotting_positions(int(args[0]), alpha=args[1])[int(args[2])]
        elif fn in ("LHSRandom", "LHSMedian"):
            m = bf.latin_hypercube(int(args[0]), int(args[1]), seed=int(args[2]),
                                   median=fn == "LHSMedian")
            actual = m[int(args[3]), int(args[4])]
        else:  # pragma: no cover
            raise AssertionError(f"unhandled function {fn}")
        a = case["assertions"][0]
        expected = float(a["expected"])
        if a["mode"] == "equal":
            assert actual == expected
        elif a["mode"] == "abs":
            assert abs(actual - expected) <= a["tol"]
        else:
            assert abs(actual - expected) / abs(expected) <= a["tol"]


def test_transform_round_trips():
    x = bf.Distribution("LogNormal", [3, 0.5]).random(50, seed=3)
    for lam in (-0.5, 0.0, 0.7):
        assert np.allclose(bf.box_cox_inverse(bf.box_cox(x, lam), lam), x)
    y = bf.Distribution("Normal", [0, 2]).random(50, seed=3)
    for lam in (0.3, 1.0, 1.8):
        assert np.allclose(bf.yeo_johnson_inverse(bf.yeo_johnson(y, lam), lam), y)


def test_plotting_positions():
    pp = bf.plotting_positions(10)
    assert pp.shape == (10,)
    assert np.all(np.diff(pp) > 0)
    assert np.all((pp > 0) & (pp < 1))
    assert np.array_equal(bf.plotting_positions(10, method="hazen"),
                          bf.plotting_positions(10, alpha=0.5))
    with pytest.raises(ValueError):
        bf.plotting_positions(10, method="nope")


def test_latin_hypercube_stratified_and_seeded():
    m = bf.latin_hypercube(20, 3, seed=5)
    assert m.shape == (20, 3)
    assert np.all((m > 0) & (m < 1))
    # LHS property: exactly one sample per 1/n bin in every dimension.
    for j in range(3):
        assert sorted(np.floor(m[:, j] * 20).astype(int)) == list(range(20))
    assert np.array_equal(bf.latin_hypercube(20, 3, seed=5), m)


def test_mgbt_flags_planted_low_outliers():
    x = bf.Distribution("LogNormal", [3, 0.5]).random(40, seed=7)
    assert bf.mgbt_test(np.concatenate([x, [1.0, 2.0]])) == 2
    assert bf.mgbt_test(x) == 0

File: corehydror/tests/testthat/test-stats.R
# Stats utility wrapper tests. Numeric expectations come from the data_utility fixture
# (fixtures/data/statistics_utilities.json) routed through the PUBLIC wrappers; the rest
# are structural (round trips, shapes, argument validation) with no oracle content.

test_that("public stats wrappers reproduce the data_utility fixture oracles", {
  skip_if_not_installed("jsonlite")
  path <- file.path(system.file("fixtures", package = "corehydror"),
    "data", "statistics_utilities.json")
  spec <- jsonlite::read_json(path, simplifyVector = FALSE)
  datasets <- lapply(spec$datasets, \(d) as.double(unlist(d)))
  for (case in spec$cases) {
    args <- if (is.null(case$args)) numeric() else as.double(unlist(case$args))
    data <- if (is.null(case$dataset)) numeric() else datasets[[case$dataset]]
    actual <- switch(case[["function"]],
      MGBT = as.double(mgbt_test(data)),
      BoxCoxLambda = box_cox_lambda(data),
      BoxCoxTransform = box_cox(data, args[[1]])[[as.integer(args[[2]]) + 1L]],
      YeoJohnsonLambda = yeo_johnson_lambda(data),
      YeoJohnsonTransform = yeo_johnson(data, args[[1]])[[as.integer(args[[2]]) + 1L]],
      PlottingPosition = plotting_positions(as.integer(args[[1]]),
        alpha = args[[2]])[[as.integer(args[[3]]) + 1L]],
      LHSRandom = latin_hypercube(as.integer(args[[1]]), as.integer(args[[2]]),
        seed = as.integer(args[[3]]))[as.integer(args[[4]]) + 1L, as.integer(args[[5]]) + 1L],
      LHSMedian = latin_hypercube(as.integer(args[[1]]), as.integer(args[[2]]),
        seed = as.integer(args[[3]]), median = TRUE)[
        as.integer(args[[4]]) + 1L, as.integer(args[[5]]) + 1L]
    )
    a <- case$assertions[[1]]
    expected <- as.numeric(a$expected)
    if (identical(a$mode, "equal")) {
      expect_equal(actual, expected)
    } else if (identical(a$mode, "abs")) {
      expect_lte(abs(actual - expected), a$tol)
    } else {
      expect_lte(abs(actual - expected) / abs(expected), a$tol)
    }
  }
})

test_that("box_cox and yeo_johnson round-trip", {
  x <- dist_random(distribution("LogNormal", c(3, 0.5)), 50, seed = 3)
  for (lambda in c(-0.5, 0, 0.7)) {
    expect_equal(box_cox_inverse(box_cox(x, lambda), lambda), x)
  }
  y <- dist_random(distribution("Normal", c(0, 2)), 50, seed = 3)
  for (lambda in c(0.3, 1, 1.8)) {
    expect_equal(yeo_johnson_inverse(yeo_johnson(y, lambda), lambda), y)
  }
})

test_that("plotting_positions validates and orders", {
  pp <- plotting_positions(10)
  expect_length(pp, 10L)
  expect_true(all(diff(pp) > 0))
  expect_true(all(pp > 0 & pp < 1))
  expect_identical(plotting_positions(10, method = "hazen"),
    plotting_positions(10, alpha = 0.5))
  expect_error(plotting_positions(10, method = "nope"))
})

test_that("latin_hypercube is stratified, seeded, and shaped", {
  m <- latin_hypercube(20, 3, seed = 5)
  expect_identical(dim(m), c(20L, 3L))
  expect_true(all(m > 0 & m < 1))
  # LHS property: exactly one sample per 1/n bin in every dimension.
  for (j in 1:3) expect_identical(sort(findInterval(m[, j], seq(0, 1, by = 1 / 20))), 1:20)
  expect_identical(latin_hypercube(20, 3, seed = 5), m)
})

test_that("mgbt_test flags planted low outliers", {
  x <- dist_random(distribution("LogNormal", c(3, 0.5)), 40, seed = 7)
  expect_identical(mgbt_test(c(x, 1, 2)), 2L)
  expect_identical(mgbt_test(x), 0L)
})
